src/global_sequences_viewer.cpp

```diff
@@ -0,0 +1,169 @@
+        auto stetefulModelInstance = std::static_pointer_cast<StatefulModelInstance>(modelInstance);
+        std::string managerId = model->getName() + std::to_string(version);
+
+        auto status = registerManager(managerId, stetefulModelInstance->getSequenceManager().get());
+        if (status.getCode() != ovms::StatusCode::OK)
+            return status;
+    }
+
+    updateThreadInterval();
+    return ovms::StatusCode::OK;
+}
+
+ovms::Status GlobalSequencesViewer::retireVersions(std::shared_ptr<ovms::Model>& model, std::shared_ptr<model_versions_t> versionsToRetire) {
+    for (const auto version : *versionsToRetire) {
+        std::string managerId = model->getName() + std::to_string(version);
+        auto status = unregisterManager(managerId);
+        if (status.getCode() != ovms::StatusCode::OK)
+            return status;
+    }
+
+    updateThreadInterval();
+    return ovms::StatusCode::OK;
+}
+
+ovms::Status GlobalSequencesViewer::reloadVersions(std::shared_ptr<ovms::Model>& model,
+    std::shared_ptr<model_versions_t> versionsToReload,
+    std::shared_ptr<model_versions_t> versionsFailed) {
+    for (const auto version : *versionsToReload) {
+        if (std::count(versionsFailed->begin(), versionsFailed->end(), version))
+            continue;
+        std::string managerId = model->getName() + std::to_string(version);
+        auto status = unregisterManager(managerId);
+        if (status.getCode() != ovms::StatusCode::OK)
+            return status;
+
+        auto modelInstance = model->getModelInstanceByVersion(version);
+        auto stetefulModelInstance = std::static_pointer_cast<StatefulModelInstance>(modelInstance);
+
+        status = registerManager(managerId, stetefulModelInstance->getSequenceManager().get());
+        if (status.getCode() != ovms::StatusCode::OK)
+            return status;
+    }
+    updateThreadInterval();
+    return ovms::StatusCode::OK;
+}
+
+void GlobalSequencesViewer::updateThreadInterval() {
+    uint32_t lowestHalfTimeoutInterval = DEFAULT_SEQUENCE_TIMEOUT_SECONDS / 2;
+    for (auto const& [key, val] : registeredSequenceManagers) {
+        auto sequenceManager = val;
+        uint32_t newInterval = sequenceManager->getTimeout() / 2;
+        if (newInterval > 0 && newInterval < lowestHalfTimeoutInterval)
+            lowestHalfTimeoutInterval = newInterval;
+    }
+
+    sequenceWatcherIntervalSec = lowestHalfTimeoutInterval;
+    SPDLOG_LOGGER_DEBUG(modelmanager_logger, "Sequence watcher thread interval set to {}", sequenceWatcherIntervalSec);
+}
+
+ovms::Status GlobalSequencesViewer::registerManager(std::string managerId, SequenceManager* sequenceManager) {
+    std::unique_lock<std::mutex> viewerLock(mutex);
+    if (registeredSequenceManagers.count(managerId)) {
+        SPDLOG_LOGGER_DEBUG(sequence_manager_logger, "Sequence manager {} already exists", managerId);
+        return StatusCode::SEQUENCE_ALREADY_EXISTS;
```

Comment:
This should not happen. I think we could return INTERNAL_ERROR here and log it at error level, not debug.